tests/unit/s2n_post_handshake_send_test.c

```diff
@@ -0,0 +1,235 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <math.h>
+#include <stdlib.h>
+
+#include "error/s2n_errno.h"
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_post_handshake.h"
+#include "tls/s2n_tls.h"
+#include "utils/s2n_safety.h"
+
+static S2N_RESULT s2n_get_expected_record_count(uint32_t nst_size, uint32_t fragment_size, uint8_t tickets_to_send,
+        uint64_t *expected_record_count)
+{
+    uint32_t app_data_records = 1;
+    uint32_t records_per_nst = ceil((1.0 * nst_size) / fragment_size);
+    uint32_t nst_records = records_per_nst * tickets_to_send;
+    *expected_record_count = app_data_records + nst_records;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_get_actual_record_count(uint8_t *cur_seq_num_bytes,
+        uint64_t *last_seq_num, uint64_t *actual_record_count)
+{
+    uint64_t cur_seq_num = 0;
+    struct s2n_blob blob = { 0 };
+    struct s2n_stuffer stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&blob, cur_seq_num_bytes, S2N_TLS_SEQUENCE_NUM_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&stuffer, &blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&stuffer, S2N_TLS_SEQUENCE_NUM_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint64(&stuffer, &cur_seq_num));
+
+    *actual_record_count = cur_seq_num - *last_seq_num;
+    *last_seq_num = cur_seq_num;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_get_nst_message_size(struct s2n_connection *conn, uint32_t *size)
+{
+    DEFER_CLEANUP(struct s2n_stuffer nst_message = { 0 }, s2n_stuffer_free);
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&nst_message, 0));
+    RESULT_GUARD(s2n_tls13_server_nst_write(conn, &nst_message));
+    *size = s2n_stuffer_data_available(&nst_message);
+    EXPECT_TRUE(*size > 50);
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+
+    /* These tests require sending tickets.
+     * Setup the config to handle tickets, but don't send any by default.
+     */
+    uint8_t ticket_key_name[16] = "key name";
+    uint8_t ticket_key[] = "key data";
+    uint64_t current_time = 0;
+    EXPECT_SUCCESS(config->wall_clock(config->sys_clock_ctx, &current_time));
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(config, 1));
+    EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(config, ticket_key_name, sizeof(ticket_key_name),
+            ticket_key, sizeof(ticket_key), current_time / ONE_SEC_IN_NANOS));
+    config->initial_tickets_to_send = 0;
+
+    const uint32_t fragment_sizes[] = {
```

Comment:
I thought fragment sizes couldn't go below 5 bytes now. Why have this in the test cases?

Reply:
Just for thoroughness. Fragment sizes currently can't go below 5 bytes in production because there's no way to set a smaller fragment size (the send buffer can only force the fragment size down to 5, and the MFL extension can only force it down to 512) but that's not really a hard rule / requirement. I'd rather know we can handle all fragment sizes that get a fun surprise if we ever add a mechanism to set even smaller fragment sizes.